private/bufpkg/bufplugin/bufplugindocker/docker.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufplugindocker
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"net"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufplugin/bufpluginconfig"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stringid"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type Client interface {
+	// Build creates a Docker image for the plugin using the Dockerfile.plugin and plugin config.
+	Build(ctx context.Context, dockerfile io.Reader, config *bufpluginconfig.Config, params BuildParams) (*BuildResponse, error)
+	// Push the Docker image to the remote registry.
+	Push(ctx context.Context, image string, auth RegistryAuth) (*PushResponse, error)
```

Comment:
Now that I see how the `RegistryAuth` is used for now, we might want to make this an optional parameter with a `PushOption` interface.

```go
// PushOption is an option for Push.
type PushOption interface func(*pushOptions)

// WithRegistryAuthorizer returns a PushOption that configures registry authentication.
func WithRegistryAuthorizer(registryAuthorizer RegistryAuthorizer) PushOption {
	return func(pushOptions *pushOptions) {
		pushOptions.registryAuthorizer = registryAuthorizer
	}
}
```

It also looks like `FromHeader` isn't used - do we need to expose that on the `RegistryAuth` type that you have here? If not, we could start with a simple interface like this:

```go
// RegistryAuthorizer implements the Docker API authorization semantics.
type RegistryAuthorizer interface {
  // Header returns the authorization information as a base64 encoded JSON object.
  // This is suitable for passing to the Docker API as the X-Registry-Auth header.
  Header() string
}
```

This also means we could construct the `Header` value upfront rather than handling it all in the `ToHeader` call. If there's ever a reason why we need to call `ToHeader` multiple times, the value will already be computed for us.

```go
// bufplugindocker.go
func NewRegistryAuthorizer(...) (RegistryAuthorizer, error) {
  return newRegistryAuthorizer(...)
}
```

```go
// registry_authorizer.go
type registryAuthorizer struct {
  header string
}

func newRegistryAuthorizer(...) (*registryAuthorizer, error) {
  ...
  return &registryAuthorizer{
    header: header,
  }
}

func (r *registryAuthorizer) Header() string {
  return r.header
}
```

But I'll admit that a lot of this is splitting hairs - the single `Header` method doesn't feel all that great either (especially with Go's interface naming guidelines). My primary goal is to outline the general approach we take with thin interface definitions that are implemented by a single un-exported struct.

Reply:
> Now that I see how the `RegistryAuth` is used for now, we might want to make this an optional parameter with a `PushOption` interface.

It is actually required - https://docs.docker.com/engine/api/v1.41/#operation/ImagePush. Even if you don't populate the auth struct, you have to at least send across something otherwise the client will fail to push.

> It also looks like `FromHeader` isn't used - do we need to expose that on the `RegistryAuth` type that you have here?

I added FromHeader really just to verify marshaling to/from headers. We won't write a Docker engine implementation any time soon though so I can un-export that method or move it to test code if you'd like.

> This also means we could construct the `Header` value upfront rather than handling it all in the `ToHeader` call. If there's ever a reason why we need to call `ToHeader` multiple times, the value will already be computed for us.

I think we could add caching to this if needed. Happy to refactor as proposed but I think this won't really need to have alternative implementations so felt simpler this way.

Reply:
Sounds good - if we keep it in its current state, let's update this to a pointer `*RegistryAuth`. We do this elsewhere, and the methods you defined on the type specify pointer receivers, so this actually matches what you've implemented there.